pkg/nfd-master/nfd-master.go

```diff
@@ -346,6 +363,84 @@ func removeLabels(n *api.Node, labelNames []string) {
 	}
 }
 
+// pruneExtResources forms a patch which removes redundant ext resources from
+// node status capacity and allocatable arrays
+// returns true if node was patched, false otherwise (and error)
+func pruneExtResources(nodeName string, labels Labels, extResources []string, oldResources []string, helper apihelper.APIHelpers) (bool, error) {
+	var err error
+	sep := ""
+	removed := []string{}
+
+	// map of ext resources for easy and fast searching
+	extResourceMap := make(map[string]string)
+	for _, name := range extResources {
+		extResourceMap[name] = ""
+	}
+
+	patch := `{"status": { "capacity": { `
+	for _, key := range oldResources {
+		// check if the related label is gone, or the ext resource requirement is gone
+		_, extResExists := extResourceMap[key]
+		if _, labelExists := labels[key]; !labelExists || !extResExists {
+			patch += sep + fmt.Sprintf(`"%s":null`, key)
+			removed = append(removed, key)
+			sep = ","
+		}
+	}
+	sep = ""
+	patch += `}, "allocatable": { `
+	for _, key := range removed {
+		patch += sep + fmt.Sprintf(`"%s":null`, key)
+		sep = ","
+	}
+	if sep == "," {
+		patch += "}}}"
+		cli, err := helper.GetClient()
+		if err != nil {
+			return false, err
+		}
+
+		err = helper.PatchStatus(cli, nodeName, []byte(patch))
+		if err == nil {
+			return true, nil
+		}
+		stderrLogger.Println(err.Error())
+	}
+	return false, err
+}
+
+// addExtResources forms a patch which adds ext resources to node.
+// returns true if node was patched, false otherwise (and error)
+func addExtResources(nodeName string, labels Labels, extResources []string, helper apihelper.APIHelpers) (bool, error) {
+	var err error
+	patch := `{"status": { "capacity": { `
+	sep := ""
+
+	for _, key := range extResources {
+		// check that there is such a label
+		if _, ok := labels[key]; ok {
+			// check that the label has a numeric int value
+			if _, err := strconv.Atoi(labels[key]); err == nil {
+				patch += sep + fmt.Sprintf(`"%s": "%s"`, key, labels[key])
+				sep = ","
+			}
+		}
+	}
+	if sep == "," {
+		patch += "}}}"
+		cli, err := helper.GetClient()
+		if err != nil {
+			return false, err
+		}
+		err = helper.PatchStatus(cli, nodeName, []byte(patch))
```

Comment:
Note: unlike pruning, this does not avoid patching even if the ext resources stay the same all the time. I do not know about the actual implementation of CoreV1().Nodes().PatchStatus, how quick it is figuring out that the diff is zero.

It could be possible to check in this function whether the resource already exists based on annotations, but then the annotation should store also the value of the extended resource capacity. Then with this check the unnecessary calling of PatchStatus could be avoided, when the resource hasn't really changed.